ql/experimental/barrieroption/mcdoublebarrierengine.cpp

```diff
@@ -0,0 +1,110 @@
+*/
+
+#include <ql/experimental/barrieroption/mcdoublebarrierengine.hpp>
+
+namespace QuantLib {
+
+    DoubleBarrierPathPricer::DoubleBarrierPathPricer(
+        DoubleBarrier::Type barrierType,
+        Real barrierLow,
+        Real barrierHigh,
+        Real rebate,
+        Option::Type type,
+        Real strike,
+        const std::vector<DiscountFactor>& discounts,
+        const ext::shared_ptr<StochasticProcess1D>& diffProcess,
+        const PseudoRandom::ursg_type& sequenceGen)
+        : barrierType_(barrierType), barrierLow_(barrierLow), barrierHigh_(barrierHigh),
+          rebate_(rebate), diffProcess_(diffProcess),
+          sequenceGen_(sequenceGen), payoff_(type, strike),
+          discounts_(discounts) {
+        QL_REQUIRE(strike>=0.0,
+                   "strike less than zero not allowed");
+        QL_REQUIRE(barrierLow>0.0,
+                   "low barrier less/equal zero not allowed");
+        QL_REQUIRE(barrierHigh>0.0,
+                   "high barrier less/equal zero not allowed");
+    }
+
+    Real DoubleBarrierPathPricer::operator()(const Path& path) const {
+        static Size null = Null<Size>();
+        Size n = path.length();
+        QL_REQUIRE(n>1, "the path cannot be empty");
+
+        bool isOptionActive = false;
+        Size knockNode = null;
+        Real asset_price = path.front();
+        Real new_asset_price;
+        TimeGrid timeGrid = path.timeGrid();
+        std::vector<Real> u = sequenceGen_.nextSequence().value;
+        Size i;
+
+        switch (barrierType_) {
+            case DoubleBarrier::KnockOut:
+                isOptionActive = true;
+                for (i = 0; i < n-1; i++) {
+                    new_asset_price = path[i + 1];
+
+                    if (new_asset_price >= barrierHigh_ || new_asset_price <= barrierLow_){
+                        isOptionActive = false;
+                        if (knockNode == null)
+                            knockNode = i+1;
+                        asset_price = path.back();
+                        break;
+                    }
+
+                    asset_price = new_asset_price;
+                }
+                break;
+            case DoubleBarrier::KnockIn:
+                isOptionActive = false;
+                for (i = 0; i < n-1; i++) {
+                    new_asset_price = path[i + 1];
+
+                    if (new_asset_price >= barrierHigh_ || new_asset_price <= barrierLow_){
+                        isOptionActive = true;
+                        if (knockNode == null)
+                            knockNode = i+1;
+                        asset_price = path.back();
+                        break;
+                    }
+
+                    asset_price = new_asset_price;
+                }
+                break;
+            default:
+                QL_FAIL("unknown barrier type");
+        }
+
+        if (isOptionActive) {
+            return payoff_(asset_price) * discounts_.back();
```

Comment:
I think you can use `path.back()` here and avoid keeping track of `asset_price` in the loops.

Reply:
Done